lib/http_v4_signature.c

```diff
@@ -0,0 +1,310 @@
+  tmp = strchr(provider, ':');
+  if(tmp) {
+    for(i = 0; provider != tmp; ++provider, ++i) {
+      if(i > PROVIDER_MAX_L)
+        return CURLE_FAILED_INIT;
+      up_provider[i] = (char)toupper(*provider);
+      low_provider0[i] = (char)tolower(*provider);
+    }
+    up_provider[i] = 0;
+    low_provider0[i] = 0;
+    ++provider;
+    for(i = 0; *provider; ++provider, ++i) {
+      if(i > PROVIDER_MAX_L)
+        return CURLE_FAILED_INIT;
+      low_provider[i] = (char)tolower(*provider);
+      mid_provider[i] = i ? low_provider[i] : (char)toupper(*provider);
+    }
+    low_provider[i] = 0;
+    mid_provider[i] = 0;
+  }
+  else if(strlen(provider) <= PROVIDER_MAX_L) {
+    for(i = 0; provider[i]; ++i) {
+      low_provider0[i] = (char)tolower(provider[i]);
+      low_provider[i] = (char)tolower(provider[i]);
+      up_provider[i] = (char)toupper(low_provider[i]);
+      mid_provider[i] = i ? low_provider[i] : up_provider[i];
+    }
+    low_provider0[i] = 0;
+    low_provider[i] = 0;
+    up_provider[i] = 0;
+    mid_provider[i] = 0;
+  }
+  else {
+    return CURLE_FAILED_INIT;
+  }
+
+  time(&rawtime);
+
+  ret = Curl_gmtime(rawtime, &info);
+  if(ret != CURLE_OK) {
+    return ret;
+  }
+
+  if(!strftime(date_iso, 256, "%Y%m%dT%H%M%SZ", &info)) {
+    return CURLE_OUT_OF_MEMORY;
+  }
+
+  memcpy(date, date_iso, 8);
+  date[8] = 0;
+  api_type = strdup(surl);
+  if(!api_type)
+    return CURLE_OUT_OF_MEMORY;
+
+  tmp = strchr(api_type, '.');
+  if(!tmp)
+    return CURLE_FAILED_INIT;
+  *tmp = 0;
+
+  region = strdup(strchr(surl, '.') + 1);
+  if(!region) {
+    ret = CURLE_OUT_OF_MEMORY;
+    goto free_all;
+  }
+
+  tmp = strchr(region, '.');
+  if(!tmp) {
+    ret = CURLE_FAILED_INIT;
+    goto free_all;
+  }
+  *tmp = 0;
+
+  uri = strchr(surl, '/');
+  host = strdup(surl);
+  if(!host) {
+    ret = CURLE_OUT_OF_MEMORY;
+    goto free_all;
+  }
+
+  query_url = strchr(host, '?');
+  if(query_url)
+    *query_url = 0;
+
+  tmp = strchr(host, '/');
+  if(tmp)
+    *tmp = 0;
+
+  curl_msprintf(request_type, "%s4_request", low_provider0);
+  cred_scope = curl_maprintf("%s/%s/%s/%s", date, region, api_type,
+                             request_type);
+  if(content_type) {
+    canonical_hdr = curl_maprintf(
+      "content-type:%s\n"
+      "host:%s\n"
+      "x-%s-date:%s\n", content_type, host, low_provider, date_iso);
+    signed_headers = curl_maprintf("content-type;host;x-%s-date",
+                                   low_provider);
+  }
+  else if(query_url) {
+    query_url++;
+    canonical_hdr = curl_maprintf(
+      "host:%s\n"
+      "x-%s-date:%s\n", host, low_provider, date_iso);
+    signed_headers = curl_maprintf("host;x-%s-date", low_provider);
+  }
+  else {
+    ret = CURLE_FAILED_INIT;
+    goto free_all;
+  }
+
+
+  Curl_sha256it(sha_d, (const unsigned char *)post_data, strlen(post_data));
+  for(i = 0; i < 32; ++i) {
+    curl_msprintf(sha_str + (i * 2), "%02x", sha_d[i]);
+  }
+  sha_str[64] = 0;
```

Comment:
This is superfluous, right? The loop above outputs the terminating null already?